src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt

```diff
@@ -28,17 +33,27 @@ import org.koin.ktor.ext.inject
  * API to create organization in IAM.
  */
 fun Route.createOrganizationApi() {
-    val service: OrganizationsService by inject()
+    val organizationService: OrganizationsService by inject()
+    val passcodeRepo: PasscodeRepo by inject()
+    val appConfig: AppConfig by inject()
     val gson: Gson by inject()
 
     post("/organizations") {
-        val request = call.receive<CreateOrganizationRequest>().validate()
-        val passcodeStr = call.principal<ApiPrincipal>()?.tokenCredential?.value
-        val (organization, tokenResponse) = service.createOrganization(
-            request.name,
-            description = request.description ?: "",
-            rootUser = request.rootUser,
-            passcodeStr = passcodeStr
+        val secretKey = Constants.SECRET_PREFIX + appConfig.app.secretKey
+        val passcodeStr = call.principal<ApiPrincipal>()?.tokenCredential?.value!!
+
+        // If secretKey is provided, then fetch use request body else use metadata from passcode table
+        val request = if (secretKey == passcodeStr)
+            call.receive<CreateOrganizationRequest>()
+        else {
+            val passcode = passcodeRepo.getValidPasscode(passcodeStr, VerifyEmailRequest.Purpose.signup)
```

Comment:
If the request has an invalid passcodeStr then this method will throw exception. Is it handled properly by throwing bad request exception to customer ?

Reply:
All the validation regarding valid `passcode` is done in authentication level.